### PR TITLE
Render timeline separator for late event groups

### PR DESCRIPTION
**Pull request by t3chguy**
For https://github.com/vector-im/wat-internal/issues/42

image

---
Here's what your changelog entry will look like:

## ✨ Features
 * Render timeline separator for late event groups ([\#11739](https://github.com/matrix-org/matrix-react-sdk/pull/11739)).

### CODE REVIEW

**Review by nadonomy — CHANGES_REQUESTED** [review]
@t3chguy This looks great! Seeing it in chorus with the tooltips, I think we can make the datestamp label way more verbose, which in turn makes the timeline a lot more legible when scanning.

So simply:

> Originally sent {{ ... }}

Instead of:

> Recovered messages originally sent at {{...}}

Sorry we could've caught that at design time, but hopefully a trivial tweak.

**Review by nadonomy — APPROVED** [review]
LGTM

File: src/components/views/messages/TimelineSeparator.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+export const enum SeparatorKind {
+    None,
+    Date,
+    LateEvent,
+}
+
+const TimelineSeparator: React.FC<Props> = ({ label, children }) => {
+    // ARIA treats <hr/>s as separators, here we abuse them slightly so manually treat this entire thing as one
```
Comment (germain-gg) [inline src/components/views/messages/TimelineSeparator.tsx:30]: Sounds like we could and probably should use pseudo elements `::before` and `::after` in place of `<hr />` here.

File: src/components/views/messages/TimelineSeparator.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    label: string;
+}
+
+export const enum SeparatorKind {
+    None,
+    Date,
+    LateEvent,
+}
+
+const TimelineSeparator: React.FC<Props> = ({ label, children }) => {
```
Comment (germain-gg) [inline src/components/views/messages/TimelineSeparator.tsx:29]: Could we add a JSDoc comment to describe this component. And similarly in `DateSeparator` to explain why we have a more specialised component.

File: src/components/structures/grouper/LateEventGrouper.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import { MatrixEvent } from "matrix-js-sdk/src/matrix";
+
+const UNSIGNED_KEY = "io.element.late_event";
+
+interface UnsignedLateEventInfo {
+    received_at: number;
+    group_id: string;
+}
+
+export function getLateEventInfo(mxEvent: MatrixEvent): UnsignedLateEventInfo | undefined {
```
Comment (germain-gg) [inline src/components/structures/grouper/LateEventGrouper.ts:26]: JSDoc comment please. a "late event" is not a concept that is widely understood I believe.

**Review by germain-gg — APPROVED** [review]
Looks good! Added a few comments, but don't consider them blocking if this needs to go in quickly

File: res/css/views/messages/_TimelineSeparator.pcss [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/*
+Copyright 2017 Vector Creations Ltd
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply res/css/views/messages/_TimelineSeparator.pcss:2]: This file is just the old one but renamed and cut in half, I didn't write any new rules or styles so it isn't new IP

File: res/css/views/messages/_TimelineSeparator.pcss [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+limitations under the License.
+*/
+
+.mx_TimelineSeparator {
+    clear: both;
+    margin: 4px 0;
+    display: flex;
+    align-items: center;
+    font: var(--cpd-font-body-md-regular);
+    color: $roomtopic-color;
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply res/css/views/messages/_TimelineSeparator.pcss:23]: These styles are not new, they're just broken out into 2 files

File: src/components/views/messages/TimelineSeparator.tsx (hunk 1 above)
Comment (t3chguy) [reply src/components/views/messages/TimelineSeparator.tsx:30]: Yup, possibly, this isn't new code, just moving home

**Pull request merged**